### PR TITLE
feat: Batch process that stores historical datasets

### LINKED ISSUE #19: Batch process that stores historical datasets
## **User Story**

As an API consumer, I need to access historical datasets so I can do service analysis and troubleshoot any issues I have with the latest dataset by referring to past versions. 

## **Acceptance Criteria**

- A batch process that compares the `producer_URL` dataset to the `latest_dataset_id` dataset and stores a new dataset when there's a change
- Batch process should include all the steps for storing historical datasets outlined in [this flowchart](https://miro.com/app/board/uXjVM86LB1s=/?moveToWidget=3458764558860791000&cot=14)
Public  Mobility Database API_ User Story Mapping - Flowchart for storing GTFS Schedule datasets (1)

## **Materials**

- [Current cronjob for storing latest datasets in the Mobility Database catalogs](https://github.com/MobilityData/mobility-database-catalogs/blob/main/.github/workflows/store_latest_datasets_cronjob.yml)

## **Considerations**

- When should we move historical datasets to a different archive? 
- We want the feed IDs to be stable and for there to be at least one piece of metadata that doesn't change after a feed is created to ensure this stability. Is there value in always storing/including the first dataset we generate?

### LINKED ISSUE #88: Drop last_update_date and creation_date fields from the GTFSDataset database entity
On the dataset, `download_date` is the only field needed. (`updated_at` is relevant only for `Feed`, since any change to a `dataset` will lead to new one).

Decision: Get rid of `last_update_date` and `creation_date`

Image

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**
A batch process that compares the `producer_URL` hash from a `Feed` to the latest persisted dataset and stores a new dataset when there's a change as well as updates entities in the database.

**Expected behavior:** 
DEV historical datasets stored [here](https://console.cloud.google.com/storage/browser/mobility-datasets;tab=objects?forceOnBucketsSortingFiltering=true&project=mobility-feeds-dev&prefix=&forceOnObjectsSortingFiltering=false).

image

Closes #19 
Closes #88 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: infra/batch/datasets/main.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+    headers = {
+        'User-Agent':
+            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) '
+            'Chrome/39.0.2171.95 Safari/537.36'
+    }
+    response = requests.get(url, headers=headers, verify=False)
+    response.raise_for_status()
+
+    content = response.content
```
Comment (jcpitre) [inline infra/batch/datasets/main.py:45]: Could the size of the file cause problems since we load it all in memory?

File: .github/workflows/datasets-batch-deployer.yml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+          scripts/replace-variables.sh -in_file infra/batch/vars.tfvars.rename_me -out_file infra/batch/pre-vars.tfvars -variables FUNCTION_BUCKET_NAME,REGION,PROJECT_ID,DEPLOYER_SERVICE_ACCOUNT,SOURCE_ZIP_PATH,SOURCE_CODE_ZIP_FILE,HTTP_FUNCTION_NAME,PUBSUB_FUNCTION_NAME,RUNTIME,HTTP_ENTRY_POINT,PUBSUB_ENTRY_POINT,AVAILABLE_MEMORY,AVAILABLE_CPU,HTTP_TIMEOUT_SECONDS,PUBSUB_TIMEOUT_SECONDS,MAX_INSTANCE_COUNT,JOB_NAME,JOB_DESCRIPTION,JOB_SCHEDULE,JOB_ATTEMPT_DEADLINE,JOB_HTTP_METHOD,PUBSUB_TOPIC_NAME,CREATE_PUBSUB_FUNCTION  
+          scripts/replace-variables.sh -in_file infra/batch/pre-vars.tfvars -out_file infra/batch/vars.tfvars -variables FUNCTION_ENV_VARIABLE -no_quotes
+          cat infra/batch/vars.tfvars
+        env:
+          FUNCTION_ENV_VARIABLE: '"BUCKET_NAME"="${{  inputs.DATASETS_BUCKET }}", "POSTGRES_USER"="${{ secrets.POSTGRES_USER }}", "POSTGRES_PASSWORD"="${{ secrets.POSTGRES_PASSWORD }}", "POSTGRES_DB"="${{ inputs.POSTGRES_DB }}", "POSTGRES_PORT"="${{ inputs.POSTGRES_PORT }}", "POSTGRES_HOST"="${{ secrets.POSTGRES_HOST }}", "PUBSUB_TOPIC_NAME"="${{ inputs.PUBSUB_TOPIC_NAME }}", "PROJECT_ID"="${{ inputs.PROJECT_ID }}"'
```
[the remark this answers is not in the document]
Comment (cka-y) [reply .github/workflows/datasets-batch-deployer.yml:187]: The replace script only uses env variables so the secret has to be written as an env variable

File: infra/batch/datasets/main.py (hunk 1 above)
Comment (cka-y) [reply infra/batch/datasets/main.py:45]: I tried to load data in chunks and compute the hash with a combination of its chunks which worked but didn't improve memory usage much.

Comment (cka-y) [reply infra/batch/datasets/main.py:45]: For the feed we currently have with the configuration I set up we do not currently have any memory problem. We should probably keep an eye on that!

File: .github/workflows/datasets-batch-deployer.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcpitre) [reply .github/workflows/datasets-batch-deployer.yml:187]: Let me rephrase: It looks to me that this line:
`echo "FUNCTION_ENV_VARIABLE=${{ secrets.FUNCTION_ENV_VARIABLE }}" >> $GITHUB_ENV`
is superseded by 

> env:
>   FUNCTION_ENV_VARIABLE: '"BUCKET_NAME"="$ ...`

Meaning it looks like it's not useful. 
Plus I did not see any FUNCTION_ENV_VARIABLE in the repo or organization secrets.
But maybe I misinterpreted.

[the remark this answers is not in the document]
Comment (cka-y) [reply .github/workflows/datasets-batch-deployer.yml:187]: Yes you’re right! I forgot to remove it. Will change it ASAP

File: infra/batch/datasets/main.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        errors += f"[ERROR]: {e}\n{error_traceback}\n"
+        print(f"Logging errors for stable id {stable_id}\n{errors}")
+        storage_client = storage.Client()
+        bucket = storage_client.get_bucket(bucket_name)
+        error_type = 'other'
+        if 'sqlalchemy' in errors:
+            error_type = 'sqlachemy'
+        elif isinstance(e, HTTPError):
+            error_type = f"http/{e.response.status_code}"
+        blob = bucket.blob(f"errors/{datetime.now().strftime('%Y%m%d')}/{error_type}/{stable_id}.log")
```
Comment (davidgamez) [inline infra/batch/datasets/main.py:145]: Having the logs in a bucket is hard to consume without an additional orchestration or tool to query the log files. I suggest removing this section here and printing the logs to the console for MVP. This will actually send the logs to `Google Cloud's operations suite` which can be consumed and queried using the `GCP Logs Explorer`. We should explore how we can format the logs in a way that queries on Logs Explorer are easy to create, in any case this can be part of different PR/issue.

File: infra/batch/datasets/main.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    storage_client = storage.Client()
+    bucket = storage_client.lookup_bucket(bucket_name)
+    if bucket is None:
+        bucket = storage_client.create_bucket(bucket_name)
+        print(f'Bucket {bucket} created.')
+    else:
+        print(f'Bucket {bucket_name} already exists.')
+
+
+def get_db_engine():
```
Comment (davidgamez) [inline infra/batch/datasets/main.py:167]: These two functions can be extracted from this files. This will allow other Cloud functions to reuse the same logic

File: infra/batch/datasets/main.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
+import base64
```
Comment (davidgamez) [inline infra/batch/datasets/main.py:1]: Functions on this script are a core functionality for keeping the dataset information up to date; let's add unit tests.

File: infra/batch/datasets/main.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        sql_statement = f"insert into gtfsdataset (id, feed_id, latest, bounding_box, note, hash, " \
+                        f"download_date, stable_id, hosted_url) " \
+                        f"select '{str(uuid.uuid4())}', feed_id, true, bounding_box, note, " \
+                        f"'{md5_file_hash}', NOW(), stable_id, '{hosted_url}' from " \
+                        f"gtfsdataset where id='{dataset_id}'"
+
+        # In case the dataset doesn't include a hash or the dataset was deleted from the bucket,
+        # update the existing entity
+        if dataset_hash is None or dataset_hash == md5_file_hash:
+            sql_statement = f"update gtfsdataset set hash='{md5_file_hash}', hosted_url='{hosted_url}' where id='{dataset_id}'"
```
Comment (davidgamez) [inline infra/batch/datasets/main.py:127]: suggestion: this is a specific use case warning us that something happened with the files or the DB integrity. I suggest logging a warning to the console to know when this use case is present.

File: infra/batch/datasets/main.py (hunk 5 above)
Comment (cka-y) [reply infra/batch/datasets/main.py:1]: Good point. This should be covered by [this issue](https://github.com/MobilityData/mobility-feed-api/issues/105).

File: infra/batch/datasets/main.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        # Set the previous version latest field to false
+        if dataset_hash is not None and dataset_hash != md5_file_hash:
+            sql_statement = f"update gtfsdataset set latest=false where id='{dataset_id}'"
+            connection.execute(text(sql_statement))
+
+        sql_statement = f"insert into gtfsdataset (id, feed_id, latest, bounding_box, note, hash, " \
+                        f"download_date, stable_id, hosted_url) " \
+                        f"select '{str(uuid.uuid4())}', feed_id, true, bounding_box, note, " \
+                        f"'{md5_file_hash}', NOW(), stable_id, '{hosted_url}' from " \
+                        f"gtfsdataset where id='{dataset_id}'"
```
[the remark this answers is not in the document]
Comment (cka-y) [reply infra/batch/datasets/main.py:122]: We agreed with @emmambd that these fields might be redundant and decided to drop them per [this issue](https://github.com/MobilityData/mobility-feed-api/issues/88) (issue is covered in the current PR see `feat_88.sql`).

File: infra/batch/datasets/main.py (hunk 6 above)
Comment (cka-y) [reply infra/batch/datasets/main.py]: Agreed and updated

File: infra/batch/datasets/main.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    headers = {
+        'User-Agent':
+            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) '
+            'Chrome/39.0.2171.95 Safari/537.36'
+    }
+    response = requests.get(url, headers=headers, verify=False)
+    response.raise_for_status()
+
+    content = response.content
+    file_md5_hash = md5(content).hexdigest()
```
[the remark this answers is not in the document]
Comment (cka-y) [reply infra/batch/datasets/main.py]: After a little digging I found out that `sha256` is more robust than `md5` even if `md5` seems faster (source: https://www.freecodecamp.org/news/md5-vs-sha-1-vs-sha-2-which-is-the-most-secure-encryption-hash-and-how-to-check-them/). I changed the hashing to `sha256` for now. We might need to keep an eye on how it performs.

File: infra/batch/datasets/main.py (hunk 3 above)
Comment (cka-y) [reply infra/batch/datasets/main.py:135]: Created [this issue](https://github.com/orgs/MobilityData/projects/64/views/1?pane=issue&itemId=39289533) to address the problem.

**Review by jcpitre — APPROVED** [review]

**Review by davidgamez — APPROVED** [review]
This is a huge work covering infrastructure and coding at the same time! 💪
